File: lab6/Java/src/request/SupplyRequest.java
package request;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Random;
import unit.Unit;

public class SupplyRequest implements Cloneable {
    int id;
    int requestUnitId;
    Unit requestUnit;
    LocalDateTime creationTime;
    ArrayList<SupplyRequestDetail> details = new ArrayList<>();
    public SupplyRequest(int id, ArrayList<SupplyRequestDetail> details, Unit requestUnit){
        this.id = id;
        this.details = details;
        this.creationTime = LocalDateTime.now();
        if(requestUnit != null){
            this.requestUnitId = requestUnit.getId();
            this.requestUnit = requestUnit;
        }
    }

    public int getId(){
        return id;
    }

    public Unit getRequestUnit(){
        return requestUnit;
    }

    public void setRequestUnit(Unit unit){
        if(unit != null){
            requestUnitId = unit.getId();
            requestUnit = unit;
        }
    }

    public ArrayList<SupplyRequestDetail> getDetails(){
        return details;
    }

    public LocalDateTime getCreationTime(){
        return creationTime;
    }

    public SupplyRequest clone(Boolean shallow) throws CloneNotSupportedException{
        // Поверхностное клонирование
        var clone = (SupplyRequest)super.clone();
        // глубокое клонирование при shallow = false
        if(!shallow){
            clone.id = new Random().nextInt(0,9999);
            clone.creationTime = LocalDateTime.now();
            clone.details = new ArrayList<>();
            for(var detail : details){
                clone.details.add((SupplyRequestDetail)detail.clone());
            }
        }
        return clone;

    }
}

File: lab6/C++/src/SupplyRequest.cpp
#include "SupplyRequest.hpp"
#include "SupplyRequestDetail.hpp"
#include "Unit.hpp"

SupplyRequest::SupplyRequest(const SupplyRequest& other) {
    m_id = rand();
    m_requestUnit = other.GetRequestUnit();
    m_requestUnitId = other.GetRequestUnitId();

    for (const auto& detail : other.m_details) {
        m_details.push_back(std::make_unique<SupplyRequestDetail>(*detail));
    }

    m_createTime = std::chrono::system_clock::now();
}

SupplyRequest::SupplyRequest(int id, const std::weak_ptr<const Unit> unit, std::vector<std::unique_ptr<SupplyRequestDetail>>& details)
    : m_id(id), m_details(std::move(details))
{
    m_createTime = std::chrono::system_clock::now();
    SetRequestUnit(unit);
}

SupplyRequest::SupplyRequest(int id, const std::weak_ptr<const Unit> unit, std::chrono::system_clock::time_point createTime) :
    m_id(id), m_createTime(createTime)
{
    SetRequestUnit(unit);
}


int SupplyRequest::GetId() const {
    return m_id;
}

int SupplyRequest::GetRequestUnitId() const {
    return m_requestUnitId;
}

const std::weak_ptr<const Unit> SupplyRequest::GetRequestUnit() const {
    return m_requestUnit;
}

void SupplyRequest::SetRequestUnit(const std::weak_ptr<const Unit> unit) {
    if (auto requestUnit = unit.lock()) {
        m_requestUnitId = requestUnit->GetId();
        m_requestUnit = unit;
    }
}

std::vector<std::unique_ptr<SupplyRequestDetail>>& SupplyRequest::GetDetails() {
    return m_details;
}

std::chrono::system_clock::time_point SupplyRequest::GetCreateTime() const {
    return m_createTime;
}

std::unique_ptr<SupplyRequest> SupplyRequest::Clone(bool shallow) const {
    return shallow ? 
        std::make_unique<SupplyRequest>(m_id, m_requestUnit, this->GetCreateTime()) :
        std::make_unique<SupplyRequest>(*this);
}